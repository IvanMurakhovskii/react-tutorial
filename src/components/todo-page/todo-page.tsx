import React, { Component } from 'react';
import TodoList from '../todo-list';
import OrderSelect from '../order-select/order-select';
import OrderEnum from '@/emums/order-enum';
import ErrorBoundry from '../error-boundry';

import { ToDoData } from '@/types'

import TodoService from '@/services/todo-service';

import AddItemForm from '../add-item-form';
import styled from '@emotion/styled';

interface State {
    todoData: ToDoData[],
    todoService: TodoService,
    order: OrderEnum
}

const AppStyle = styled.div`
    margin: 2rem auto 0 auto;
    width: 60%;
`;

export default class TodoPage extends Component<{}, State> {

    nextId: Function = () => { };

    constructor(props: any) {
        super(props);

        this.state = {
            todoData: [],
            todoService: new TodoService(),
            order: OrderEnum.ASC
        };

        this.deleteItem = this.deleteItem.bind(this);
    }

    componentDidMount() {
        const service = this.state.todoService;
        service.getAllTodos()
            .then((data) => {
                this.setState({
                    todoData: data
                });

                this.nextId = this.getNextId();
            });
    }

    deleteItem(id: number) {

        this.setState(({ todoData }) => {
            const idx = todoData.findIndex((el) => el.id === id);

            const newTodoData = [
                ...todoData.slice(0, idx),
                ...todoData.slice(idx + 1)
            ];

            this.state.todoService.saveTodos(newTodoData);

            return {
                todoData: newTodoData
            }
        });
    };

    toggleProperty(arr: ToDoData[], id: number, propName: "important" | "done") {
        const idx = arr.findIndex((el) => el.id === id);

        const oldItem = arr[idx];
        const newItem = { ...oldItem, [propName]: !oldItem[propName] };

        return [
            ...arr.slice(0, idx),
            newItem,
            ...arr.slice(idx + 1)
        ];
    }

    onToggleImportant = (id: number) => {
        this.setState(({ todoData }) => {
            return {
                todoData: this.toggleProperty(todoData, id, 'important')
            }
        });
    };

    onToggleDone = (id: number) => {
        this.setState(({ todoData }) => {
            return {
                todoData: this.toggleProperty(todoData, id, 'done')
            }
        });
    };

    createTodoItem(label: string) {
        return {
            label,
            important: false,
            done: false,
            hidden: false,
            id: this.nextId()
        }
    }

    onOrderChange = (order: OrderEnum) => {
        this.setState({ order: order })
    }

    addItem = (label: string) => {

        const newTodo = this.createTodoItem(label);

        this.setState(({ todoData }) => {
            const newTodos = [
                ...todoData,
                newTodo
            ];

            this.state.todoService.saveTodos(newTodos);

            return {
                todoData: newTodos
            };
        });
    }

    getNextId = (): Function => {
        let maxId = Math.max(...this.state.todoData.map(item => item.id), 0);

        return function nextId() {
            return ++maxId;
        }
    }

    render() {
        return (
            <AppStyle>
                <ErrorBoundry>
                    <OrderSelect
                        onOrderChange={this.onOrderChange}
                        order={this.state.order} />
                    <TodoList
                        todos={this.state.todoData}
                        order={this.state.order}
                        onDeleted={this.deleteItem}
                        onToggleImportant={this.onToggleImportant}
                        onToggleDone={this.onToggleDone} />
                    <AddItemForm addItem={this.addItem} />
                </ErrorBoundry>
            </AppStyle>
        );
    };
};