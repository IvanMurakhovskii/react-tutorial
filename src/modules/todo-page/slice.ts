import OrderEnum from "@/emums/order-enum";
import { ToDoData } from "@/types";
import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export type TodoState = {
    todoData: Array<ToDoData>,
    order: OrderEnum,
    username: string
}

export const initialState: TodoState = {
    todoData: new Array(),
    order: OrderEnum.ASC,
    username: ''
};

const todo = createSlice({
    name: "todo",
    initialState: initialState,
    reducers: {
        important: (state, action: PayloadAction<number>) => {
            const todos = state.todoData;
            const id = action.payload;
            const idx = todos.findIndex((el) => el.id === id);

            const oldItem = todos[idx];
            const newItem = { ...oldItem, important: !oldItem.important };

            state.todoData = [
                ...todos.slice(0, idx),
                newItem,
                ...todos.slice(idx + 1)
            ];
        },
        done: (state, action: PayloadAction<number>) => {
            const todos = state.todoData;
            const id = action.payload;
            const idx = todos.findIndex((el) => el.id === id);

            const oldItem = todos[idx];
            const newItem = { ...oldItem, done: !oldItem.done };

            state.todoData = [
                ...todos.slice(0, idx),
                newItem,
                ...todos.slice(idx + 1)
            ];
        },
        delete: (state, action: PayloadAction<number>) => {
            const id = action.payload;
            const idx = state.todoData.findIndex((el) => el.id === id);

            const newTodoData = [
                ...state.todoData.slice(0, idx),
                ...state.todoData.slice(idx + 1)
            ];

            state.todoData = newTodoData;
        },
        addTodo: (state, action: PayloadAction<ToDoData>) => {
            const newTodos = [
                ...state.todoData,
                action.payload
            ];

            state.todoData = newTodos;
        },
        addTodos: (state, action: PayloadAction<Array<ToDoData>>) => {
            if (action.payload !== undefined) {
                state.todoData = action.payload;
            }
        },
        changeOrder: (state, action: PayloadAction<OrderEnum>) => {
            state.order = action.payload;
        },
        setUsername: (state, action: PayloadAction<string>) => {
            state.username = action.payload;
        },
        getUsername: () => {
        },
        loadTodos: () => {
        }
    }
});

export const { actions, reducer } = todo;
