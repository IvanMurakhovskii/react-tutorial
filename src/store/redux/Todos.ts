import OrderEnum from "@/emums/order-enum";
import { ToDoData } from "@/types";
import { getUsername } from "@/utils";
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

export type  TodoState = {
    todoData: ToDoData[],
    order: OrderEnum,
    username: string,
    loading: boolean,
    hasError: boolean
}

export const initialState: TodoState = {
    todoData: [],
    order: OrderEnum.ASC,
    username: '',
    loading: false,
    hasError: false
};

export const loadUsername = createAsyncThunk(
    'username', async () => {
       return getUsername(); 
    }
);


export const todo = createSlice({
    name: "todo",
    initialState: initialState,
    reducers: {
        important: (state, action: PayloadAction<number>) => {
            const todos = state.todoData;
            const id = action.payload;
            const idx = todos.findIndex((el) => el.id === id);

            const oldItem = todos[idx];
            const newItem = { ...oldItem, important: !oldItem.important};
    
            state.todoData = [
                ...todos.slice(0, idx),
                newItem,
                ...todos.slice(idx + 1)
            ];
        },
        done: (state, action: PayloadAction<number>) => {
            const todos = state.todoData;
            const id = action.payload;
            const idx = todos.findIndex((el) => el.id === id);

            const oldItem = todos[idx];
            const newItem = { ...oldItem, done: !oldItem.done};
    
            state.todoData = [
                ...todos.slice(0, idx),
                newItem,
                ...todos.slice(idx + 1)
            ];
        },
        delete: (state, action: PayloadAction<number>) => {
            const id = action.payload;
            const idx = state.todoData.findIndex((el) => el.id === id);

            const newTodoData = [
                ...state.todoData.slice(0, idx),
                ...state.todoData.slice(idx + 1)
            ];

            state.todoData = newTodoData;
        },
        addTodo: (state, action: PayloadAction<ToDoData>) => {
            const newTodos = [
                ...state.todoData,
                action.payload
            ];

            state.todoData = newTodos;
        },
        changeOrder: (state, action: PayloadAction<OrderEnum>) => {
            state.order = action.payload;
        }
    },
    extraReducers: builder => {
        builder.addCase(loadUsername.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(loadUsername.fulfilled, (state, action: PayloadAction<string>) => {
            state.username = action.payload;
            state.loading = false;
        });
        builder.addCase(loadUsername.rejected, (state) => {
            state.loading = false;
            state.hasError = true;
        });
      },
});
