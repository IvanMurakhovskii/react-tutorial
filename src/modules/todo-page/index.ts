import { Todos } from './todo-page'

export default Todos;