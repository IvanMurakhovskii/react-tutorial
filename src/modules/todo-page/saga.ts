import { todoService } from '@/services/todo-service';
import { ToDoData } from "@/types";
import { getUsername } from '@/utils';
import { call, put, takeEvery, select } from "redux-saga/effects"
import { StoreState } from '../../store/store';
import { actions } from "./slice";


export const getTodoData = ({ todoReducer }: StoreState): Array<ToDoData> => todoReducer.todoData;

export function* loadTodos() {
    const todos = yield call(todoService.getAllTodos);
    yield put(actions.addTodos(todos));
}

export function* saveTodos() {
    const todos = yield select(getTodoData);
    yield call(todoService.saveTodos, todos);
}

export function* loadUsername() {
    const username = yield call(getUsername);
    yield put(actions.setUsername(username));
}

export function* todoSaga() {
    yield takeEvery(actions.loadTodos.type, loadTodos);
    yield takeEvery(actions.addTodo.type, saveTodos);
    yield takeEvery(actions.getUsername.type, loadUsername);
    yield takeEvery(actions.important.type, saveTodos);
    yield takeEvery(actions.done.type, saveTodos);
    yield takeEvery(actions.delete.type, saveTodos);
}