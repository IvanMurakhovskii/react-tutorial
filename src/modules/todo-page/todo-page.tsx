import React, { Component } from 'react';
import TodoList from '@/components/todo-list';
import OrderSelect from '@/components/order-select/order-select';
import ErrorBoundry from '@/components/error-boundry';

import { todoService } from '@/services/todo-service';

import AddItemForm from '@/components/add-item-form';
import styled from '@emotion/styled';
import Header from '@/components/header';
import { actions } from '@/modules/todo-page/slice'
import { connect } from 'react-redux';
import { StoreState } from '@/store/store';
import { isUserAuth } from '@/utils/LogInUtil';

const TodosStyle = styled.div`
    margin: 2rem auto 0 auto;
    width: 60%;
`;

const mapStateToProps = ({ todoReducer }: StoreState) => ({
    ...todoReducer
});

const mapDispatchToProps = {
    addTodo: actions.addTodo,
    loadTodos: actions.loadTodos,
    changeOrder: actions.changeOrder,
    delete: actions.delete,
    done: actions.done,
    important: actions.important,
    getUseraname: actions.getUsername
}

export type Props = ReturnType<typeof mapStateToProps> &
    typeof mapDispatchToProps;

class TodoPage extends Component<Props, {}> {

    nextId: Function = () => { };

    constructor(props: Props) {
        super(props);
    }

    componentDidMount() {
        this.props.loadTodos();
        this.props.getUseraname();

        this.nextId = todoService.getNextId(this.props.todoData);
    }

    addItem = (label: string) => {
        const newTodo = todoService.createTodoItem(label, this.nextId());
        this.props.addTodo(newTodo);
    }

    render() {
        return (
            <TodosStyle>
                <Header username={this.props.username} isAuth={isUserAuth(this.props.username)} />
                <ErrorBoundry>
                    <OrderSelect
                        onOrderChange={this.props.changeOrder}
                        order={this.props.order} />
                    <TodoList
                        todos={this.props.todoData}
                        order={this.props.order}
                        deleteItem={this.props.delete}
                        toggleImportant={this.props.important}
                        toggleDone={this.props.done} />
                    <AddItemForm addItem={this.addItem} />
                </ErrorBoundry>
            </TodosStyle>
        );
    };
};

export const Todos = connect(mapStateToProps, mapDispatchToProps)(TodoPage);