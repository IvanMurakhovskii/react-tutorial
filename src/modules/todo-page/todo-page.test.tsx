import React, { Component } from 'react';
import { Provider } from "react-redux";

import configureStore, { MockStoreEnhanced } from 'redux-mock-store';
import Todos from "./index";
import { TodoState } from '@/modules/todo-page/slice'
import { mount, ReactWrapper } from "enzyme";
import OrderEnum from '@/emums/order-enum';

const mockStore = configureStore();

export const initialState: TodoState = {
    todoData: [],
    order: OrderEnum.ASC,
    username: 'username'
};

describe("todo-page", () => {
    let store: MockStoreEnhanced<unknown, {}>;
    let wrapper: ReactWrapper<TodoState, Readonly<{}>, Component<{}, {}, TodoState>>;

    beforeEach(() => {
        store = mockStore(initialState);

        store.dispatch = jest.fn();

        wrapper = mount(
            <Provider store={store}>
                <Todos />
            </Provider>
        );

    });

    it('should render with given state from Redux store', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should dispatch an action when component did mount', () => {
        expect(store.dispatch).toHaveBeenCalledTimes(2);
    });
});