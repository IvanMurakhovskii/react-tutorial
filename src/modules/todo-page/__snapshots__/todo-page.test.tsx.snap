// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo-page should render with given state from Redux store 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "payload": undefined,
              "type": "todo/loadTodos",
            },
          ],
          Array [
            Object {
              "payload": undefined,
              "type": "todo/getUsername",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TodoPage)>
    <TodoPage
      addTodo={[Function]}
      changeOrder={[Function]}
      delete={[Function]}
      done={[Function]}
      getUseraname={[Function]}
      important={[Function]}
      loadTodos={[Function]}
    >
      <Styled(div)>
        <div
          className="css-1sebohw"
        >
          <Header
            isAuth={false}
          >
            <WithStyles(ForwardRef(AppBar))
              position="sticky"
            >
              <ForwardRef(AppBar)
                classes={
                  Object {
                    "colorDefault": "MuiAppBar-colorDefault",
                    "colorInherit": "MuiAppBar-colorInherit",
                    "colorPrimary": "MuiAppBar-colorPrimary",
                    "colorSecondary": "MuiAppBar-colorSecondary",
                    "colorTransparent": "MuiAppBar-colorTransparent",
                    "positionAbsolute": "MuiAppBar-positionAbsolute",
                    "positionFixed": "MuiAppBar-positionFixed",
                    "positionRelative": "MuiAppBar-positionRelative",
                    "positionStatic": "MuiAppBar-positionStatic",
                    "positionSticky": "MuiAppBar-positionSticky",
                    "root": "MuiAppBar-root",
                  }
                }
                position="sticky"
              >
                <WithStyles(ForwardRef(Paper))
                  className="MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary"
                  component="header"
                  elevation={4}
                  square={true}
                >
                  <ForwardRef(Paper)
                    className="MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary"
                    classes={
                      Object {
                        "elevation0": "MuiPaper-elevation0",
                        "elevation1": "MuiPaper-elevation1",
                        "elevation10": "MuiPaper-elevation10",
                        "elevation11": "MuiPaper-elevation11",
                        "elevation12": "MuiPaper-elevation12",
                        "elevation13": "MuiPaper-elevation13",
                        "elevation14": "MuiPaper-elevation14",
                        "elevation15": "MuiPaper-elevation15",
                        "elevation16": "MuiPaper-elevation16",
                        "elevation17": "MuiPaper-elevation17",
                        "elevation18": "MuiPaper-elevation18",
                        "elevation19": "MuiPaper-elevation19",
                        "elevation2": "MuiPaper-elevation2",
                        "elevation20": "MuiPaper-elevation20",
                        "elevation21": "MuiPaper-elevation21",
                        "elevation22": "MuiPaper-elevation22",
                        "elevation23": "MuiPaper-elevation23",
                        "elevation24": "MuiPaper-elevation24",
                        "elevation3": "MuiPaper-elevation3",
                        "elevation4": "MuiPaper-elevation4",
                        "elevation5": "MuiPaper-elevation5",
                        "elevation6": "MuiPaper-elevation6",
                        "elevation7": "MuiPaper-elevation7",
                        "elevation8": "MuiPaper-elevation8",
                        "elevation9": "MuiPaper-elevation9",
                        "outlined": "MuiPaper-outlined",
                        "root": "MuiPaper-root",
                        "rounded": "MuiPaper-rounded",
                      }
                    }
                    component="header"
                    elevation={4}
                    square={true}
                  >
                    <header
                      className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
                    >
                      <WithStyles(ForwardRef(Toolbar))>
                        <ForwardRef(Toolbar)
                          classes={
                            Object {
                              "dense": "MuiToolbar-dense",
                              "gutters": "MuiToolbar-gutters",
                              "regular": "MuiToolbar-regular",
                              "root": "MuiToolbar-root",
                            }
                          }
                        >
                          <div
                            className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
                          >
                            <WithStyles(ForwardRef(Typography))
                              className="makeStyles-title-1"
                              variant="h6"
                            >
                              <ForwardRef(Typography)
                                className="makeStyles-title-1"
                                classes={
                                  Object {
                                    "alignCenter": "MuiTypography-alignCenter",
                                    "alignJustify": "MuiTypography-alignJustify",
                                    "alignLeft": "MuiTypography-alignLeft",
                                    "alignRight": "MuiTypography-alignRight",
                                    "body1": "MuiTypography-body1",
                                    "body2": "MuiTypography-body2",
                                    "button": "MuiTypography-button",
                                    "caption": "MuiTypography-caption",
                                    "colorError": "MuiTypography-colorError",
                                    "colorInherit": "MuiTypography-colorInherit",
                                    "colorPrimary": "MuiTypography-colorPrimary",
                                    "colorSecondary": "MuiTypography-colorSecondary",
                                    "colorTextPrimary": "MuiTypography-colorTextPrimary",
                                    "colorTextSecondary": "MuiTypography-colorTextSecondary",
                                    "displayBlock": "MuiTypography-displayBlock",
                                    "displayInline": "MuiTypography-displayInline",
                                    "gutterBottom": "MuiTypography-gutterBottom",
                                    "h1": "MuiTypography-h1",
                                    "h2": "MuiTypography-h2",
                                    "h3": "MuiTypography-h3",
                                    "h4": "MuiTypography-h4",
                                    "h5": "MuiTypography-h5",
                                    "h6": "MuiTypography-h6",
                                    "noWrap": "MuiTypography-noWrap",
                                    "overline": "MuiTypography-overline",
                                    "paragraph": "MuiTypography-paragraph",
                                    "root": "MuiTypography-root",
                                    "srOnly": "MuiTypography-srOnly",
                                    "subtitle1": "MuiTypography-subtitle1",
                                    "subtitle2": "MuiTypography-subtitle2",
                                  }
                                }
                                variant="h6"
                              >
                                <h6
                                  className="MuiTypography-root makeStyles-title-1 MuiTypography-h6"
                                >
                                  Todo App
                                </h6>
                              </ForwardRef(Typography)>
                            </WithStyles(ForwardRef(Typography))>
                            <WithStyles(ForwardRef(Button))
                              className="makeStyles-label-2"
                              color="secondary"
                              onClick={[Function]}
                              size="small"
                              variant="contained"
                            >
                              <ForwardRef(Button)
                                className="makeStyles-label-2"
                                classes={
                                  Object {
                                    "colorInherit": "MuiButton-colorInherit",
                                    "contained": "MuiButton-contained",
                                    "containedPrimary": "MuiButton-containedPrimary",
                                    "containedSecondary": "MuiButton-containedSecondary",
                                    "containedSizeLarge": "MuiButton-containedSizeLarge",
                                    "containedSizeSmall": "MuiButton-containedSizeSmall",
                                    "disableElevation": "MuiButton-disableElevation",
                                    "disabled": "Mui-disabled",
                                    "endIcon": "MuiButton-endIcon",
                                    "focusVisible": "Mui-focusVisible",
                                    "fullWidth": "MuiButton-fullWidth",
                                    "iconSizeLarge": "MuiButton-iconSizeLarge",
                                    "iconSizeMedium": "MuiButton-iconSizeMedium",
                                    "iconSizeSmall": "MuiButton-iconSizeSmall",
                                    "label": "MuiButton-label",
                                    "outlined": "MuiButton-outlined",
                                    "outlinedPrimary": "MuiButton-outlinedPrimary",
                                    "outlinedSecondary": "MuiButton-outlinedSecondary",
                                    "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                                    "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                                    "root": "MuiButton-root",
                                    "sizeLarge": "MuiButton-sizeLarge",
                                    "sizeSmall": "MuiButton-sizeSmall",
                                    "startIcon": "MuiButton-startIcon",
                                    "text": "MuiButton-text",
                                    "textPrimary": "MuiButton-textPrimary",
                                    "textSecondary": "MuiButton-textSecondary",
                                    "textSizeLarge": "MuiButton-textSizeLarge",
                                    "textSizeSmall": "MuiButton-textSizeSmall",
                                  }
                                }
                                color="secondary"
                                onClick={[Function]}
                                size="small"
                                variant="contained"
                              >
                                <WithStyles(ForwardRef(ButtonBase))
                                  className="MuiButton-root MuiButton-contained makeStyles-label-2 MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                                  component="button"
                                  disabled={false}
                                  focusRipple={true}
                                  focusVisibleClassName="Mui-focusVisible"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <ForwardRef(ButtonBase)
                                    className="MuiButton-root MuiButton-contained makeStyles-label-2 MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                                    classes={
                                      Object {
                                        "disabled": "Mui-disabled",
                                        "focusVisible": "Mui-focusVisible",
                                        "root": "MuiButtonBase-root",
                                      }
                                    }
                                    component="button"
                                    disabled={false}
                                    focusRipple={true}
                                    focusVisibleClassName="Mui-focusVisible"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    <button
                                      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-label-2 MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                                      disabled={false}
                                      onBlur={[Function]}
                                      onClick={[Function]}
                                      onDragLeave={[Function]}
                                      onFocus={[Function]}
                                      onKeyDown={[Function]}
                                      onKeyUp={[Function]}
                                      onMouseDown={[Function]}
                                      onMouseLeave={[Function]}
                                      onMouseUp={[Function]}
                                      onTouchEnd={[Function]}
                                      onTouchMove={[Function]}
                                      onTouchStart={[Function]}
                                      tabIndex={0}
                                      type="button"
                                    >
                                      <span
                                        className="MuiButton-label"
                                      >
                                        LogIn
                                      </span>
                                      <WithStyles(memo)
                                        center={false}
                                      >
                                        <ForwardRef(TouchRipple)
                                          center={false}
                                          classes={
                                            Object {
                                              "child": "MuiTouchRipple-child",
                                              "childLeaving": "MuiTouchRipple-childLeaving",
                                              "childPulsate": "MuiTouchRipple-childPulsate",
                                              "ripple": "MuiTouchRipple-ripple",
                                              "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                                              "rippleVisible": "MuiTouchRipple-rippleVisible",
                                              "root": "MuiTouchRipple-root",
                                            }
                                          }
                                        >
                                          <span
                                            className="MuiTouchRipple-root"
                                          >
                                            <TransitionGroup
                                              childFactory={[Function]}
                                              component={null}
                                              exit={true}
                                            />
                                          </span>
                                        </ForwardRef(TouchRipple)>
                                      </WithStyles(memo)>
                                    </button>
                                  </ForwardRef(ButtonBase)>
                                </WithStyles(ForwardRef(ButtonBase))>
                              </ForwardRef(Button)>
                            </WithStyles(ForwardRef(Button))>
                          </div>
                        </ForwardRef(Toolbar)>
                      </WithStyles(ForwardRef(Toolbar))>
                    </header>
                  </ForwardRef(Paper)>
                </WithStyles(ForwardRef(Paper))>
              </ForwardRef(AppBar)>
            </WithStyles(ForwardRef(AppBar))>
          </Header>
          <ErrorBoundry>
            <h1>
              Something wrong!
            </h1>
          </ErrorBoundry>
        </div>
      </Styled(div)>
    </TodoPage>
  </Connect(TodoPage)>
</Provider>
`;
